lib/collection/src/operations/snapshot_storage_ops.rs

```diff
@@ -113,6 +114,10 @@ pub async fn multipart_upload(
     let mut reader = BufReader::new(file);
     let mut buffer = vec![0u8; chunk_size];
 
+    // Initialize CpuBudget to manage concurrency
+    let cpu_budget = CpuBudget::default();
+    let max_concurrency = cpu_budget.available_cpu_budget();
+
```

Comment:
I will defer to repo maintainers here but I'm curious whether it is better to use available RAM here as a constraining factor instead of the number of available CPUs.

For example, a hard cap of 2GiB can be put on the uploading process (or alternatively, compute this threshold based on the available RAM) and then infer the concurrency limit based on the chunk size. A care would have to be taken not to exceed the limit when passing a `buffer` to the `write` method:
```rust
let mut copy = std::mem::take(&mut buffer);
copy.truncate(bytes_read);
write.put(copy.into()); // 1. write.write() is sync but a worker thread is spawned internally.
write.wait_for_capacity(...).await?;
buffer.resize(chunk_size, 0u8);
```

Reply:
Makes sense!

Reply:
I am thinking of @generall /  @timvisee review on the above suggestion. But still the `wait_for_capacity`  expects for max concurrency param.